src/main/java/com/spikes2212/path/PurePursuitController.java

```diff
@@ -0,0 +1,128 @@
+        this.lookaheadDistance = lookaheadDistance;
+        this.robotWidth = robotWidth;
+    }
+
+    public OdometryHandler getHandler() {
+        return handler;
+    }
+
+    public void setHandler(OdometryHandler handler) {
+        this.handler = handler;
+    }
+
+    public Path getPath() {
+        return path;
+    }
+
+    public void setPath(Path path) {
+        this.path = path;
+    }
+
+    public double getLookaheadDistance() {
+        return lookaheadDistance;
+    }
+
+    public void setLookaheadDistance(double lookaheadDistance) {
+        this.lookaheadDistance = lookaheadDistance;
+    }
+
+    private Waypoint closestPoint() {
+        Waypoint robot = handler.getWaypoint();
+        double minDistance = Double.POSITIVE_INFINITY, distance;
+        int minIndex = lastClosestIndex;
+        for (int i = lastClosestIndex; i < path.getPoints().size(); i++) {
+            if ((distance = path.getPoints().get(i).distance(robot)) < minDistance) {
+                minIndex = i;
+                minDistance = distance;
+            }
+        }
+        lastClosestIndex = minIndex;
+        return path.getPoints().get(minIndex);
+    }
+
+    private Waypoint lookaheadPoint() throws LookaheadPointNotFoundException {
+        Waypoint robot = handler.getWaypoint();
+        int lookaheadIndex = lastLookaheadIndex;
+        for (int i = lastLookaheadIndex; i < path.getPoints().size(); i++) {
+            Waypoint d = new Waypoint(path.getPoints().get(i + 1).getX() - path.getPoints().get(i).getX()
+                    , path.getPoints().get(i + 1).getY() - path.getPoints().get(i).getY());
+            Waypoint f = new Waypoint(path.getPoints().get(i).getX() - handler.getX()
+                    , path.getPoints().get(i).getY() - handler.getY());
+            double a = d.getX() * d.getX() + d.getY() * d.getY();
+            double b = 2 * (f.getX() * d.getX() + f.getY() * f.getX());
+            double c = f.getX() * f.getX() + f.getY() * f.getY();
+            double discriminant = b * b - 4 * a * c;
+            if (discriminant >= 0) {
+                discriminant = Math.sqrt(discriminant);
+                double t1 = (-b - discriminant) / (2 * a);
+                double t2 = (-b + discriminant) / (2 * a);
+                if (t1 >= 0 && t1 <= 1) {
+                    lastLookaheadIndex = i;
+                    return new Waypoint(path.getPoints().get(i).getX() + t1 * d.getX(),
+                            path.getPoints().get(i).getY() + t1 * d.getY());
+                }
+                if (t2 >= 0 && t2 <= 1) {
+                    lastLookaheadIndex = i;
+                    return new Waypoint(path.getPoints().get(i).getX() + t2 * d.getX(),
+                            path.getPoints().get(i).getY() + t2 * d.getY());
+                }
+            }
+        }
+        throw new LookaheadPointNotFoundException();
+    }
+
+    private double pathCurvature() throws LookaheadPointNotFoundException {
```

Comment:
please add documentation for all complex computations done by all functions. use the documentation for `OdometryHandler` as reference.